plugins/woocommerce/src/Blocks/Utils/ProductCollectionUtils.php

```diff
@@ -78,6 +79,87 @@ public static function remove_query_array( $queries, $key, $value ) {
 		return self::remove_empty_array_recursive( $queries );
 	}
 
+	/**
+	 * Parse WP Query's front-end context for the Product Collection block.
+	 *
+	 * The sourceData structure depends on the context type as follows:
+	 * - site:    [ ]
+	 * - order:   [ 'orderId'    => int ]
+	 * - cart:    [ 'productIds' => int[] ]
+	 * - archive: [ 'taxonomy'   => string, 'termId' => int ]
+	 * - product: [ 'productId'  => int ]
+	 *
+	 * @return array $context {
+	 *     @type string  $type        The context type. Possible values are 'site', 'order', 'cart', 'archive', 'product'.
+	 *     @type array   $sourceData  The context source data. Can be the product ID of the viewed product, the order ID of the current order, etc.
+	 * }
+	 */
+	public static function parse_frontend_location_context() {
+		global $wp_query;
+
+		// Default context.
+		// Hint: The Shop page uses the default context.
+		$type        = 'site';
+		$source_data = array();
+
+		if ( ! ( $wp_query instanceof WP_Query ) ) {
+
+			return array(
+				'type'       => $type,
+				'sourceData' => $source_data,
+			);
+		}
+
+		// As more areas are blockified, expected future contexts include:
+		// - is_checkout_pay_page().
+		// - is_view_order_page().
+		if ( is_order_received_page() ) {
+
+			$type        = 'order';
+			$source_data = array( 'orderId' => absint( $wp_query->query_vars['order-received'] ) );
+
+		} elseif ( ( is_cart() || is_checkout() ) && isset( WC()->cart ) && is_a( WC()->cart, 'WC_Cart' ) ) {
+
+			$type  = 'cart';
+			$items = array();
+			foreach ( WC()->cart->get_cart() as $cart_item ) {
+				if ( ! isset( $cart_item['product_id'] ) ) {
+					continue;
+				}
+
+				$items[] = absint( $cart_item['product_id'] );
+			}
+			$items       = array_unique( array_filter( $items ) );
```

Comment:
Just curious: Is there a possibility of duplicate product IDs in the Cart?

Reply:
Yes, @imanish003!

Consider the example where a product with Product Add-Ons might have a different footprint in cart based on the configured add-ons. Same for every other cases where the `$cart_item_data` differs from some reason. Makes sense?